### DISCUSSION TITLE
asyncpg slower than psycopg2?

### DISCUSSION THREAD

**Discussion by sagrawal-idrc** [fastapi/fastapi#7727] [Questions] [answered]
Sorry for a somewhat unrelated query, but I am unable to figure this out. Please close this issue if this is out of scope. Thanks.

Ref: https://github.com/MagicStack/asyncpg/issues/322

I am building a simple FastAPI app, without using an ORM, but using psycopg2 and asyncpg directly.

I am finding that same responses from psycopg2 are upto 30% faster than asyncpg. Queries are simple selects.

Does this mean that for simple webapps using psycopg2 is better?

**Comment by dmontagu** [comment, marked as answer]
It depends on a number of factors.

In general, for handling a single request, you would expect non-async functions to run faster than async functions because they don't need to make use of the event loop. (I can't speak to the impact of performance differences between psycopg2 and asyncpg though.)

If you are handling requests for lots of users concurrently though, asyncpg should scale better because it can make use of the event loop rather than needing to use a separate thread to handle every concurrent request.

Also, if you are handling large amounts of data in a single query, the discrepancy should quickly become smaller. Simple selects tend not to be very representative of real world queries. Not to say that you wouldn't necessarily have similar results with real queries, you'd just have to try and see. (But I would *expect* the difference to be much smaller.)

In general, if you don't expect many concurrent requests, then just using psycopg2 and sqlalchemy orm is totally fine.

**Comment by sagrawal-idrc** [comment]
Thanks for the quick response.

> If you are handling requests for lots of users concurrently though, asyncpg should scale better because it can make use of the event loop rather than needing to use a separate thread to handle every concurrent request.

Yes, I understand this.

> Also, if you are handling large amounts of data in a single query, the discrepancy should quickly become smaller. Simple selects tend not to be very representative of real world queries. Not to say that you wouldn't necessarily have similar results with real queries, you'd just have to try and see. (But I would expect the difference to be much smaller.)

Yes, I will try this out.

> In general, if you don't expect many concurrent requests, then just using psycopg2 and sqlalchemy orm is totally fine.

Thanks for taking the time to respond and the advice. Closing the issue for now.

**Comment by tiangolo** [comment]
Thanks for the help here @dmontagu ! :cake: :bowing_man: 

Thanks @sagrawal-idrc for reporting back and closing the issue :+1:
